core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -368,6 +368,27 @@ private[deploy] class Master(
       if (canCompleteRecovery) { completeRecovery() }
     }
 
+    case WorkerLatestState(workerId, executors, driverIds) =>
+      idToWorker.get(workerId) match {
+        case Some(worker) =>
+          for (exec <- executors) {
+            if (!worker.executors.exists(
+              e => e._2.application.id == exec.appId && e._2.id == exec.execId)) {
```

Comment:
style: can you use `.exists { case (_, something) => something.application.id ... }` and store it in a variable? e.g.
```
for (exec <- executors) {
  val executorMatches = worker.executors.exists { ... }
  if (!executorMatches) {
    worker.endpoint.send(...)
  }
}
```